### DISCUSSION TITLE
Overview of Mapping, Transformation, and Delivery Standards

### DISCUSSION THREAD

**Discussion by jshoughtaling** [chorus-ai/StandardsModule#87] [Announcements]
# CHoRUS Data Mapping, Transformation, and Delivery Standards

## Visual Workflow
![b2ai-transform drawio](https://github.com/chorus-ai/StandardsModule/assets/67749079/f5565fcb-ddc4-4ea2-af9f-b59ff6cba7d0)

### General Description

**Source Characterization** refers to using open-source tooling like WhiteRabbit and other internal data analysis methods to investigate and understand the data available to each data contributing site. For relational EHR data, this typically requires running a database scan or producing metadata about the contents of relevant tables. For non-relational data, characterizations will likely focus on identifying quantity (storage space, number of files, etc) and diversity (unique codes, ontology structures, etc.) of data and an overview of the metadata available that will require mapping in subsequent stages.

### Expected Outputs

- Informal overview of data available
- Rough indication of the scope of work required to map data to OMOP CDM + CHoRUS conventions

### Relevant Resources

*OFFICE HOURS*

- [WhiteRabbit](https://drive.google.com/drive/folders/1QoQsawCY-AorcpOmz5bgzZACewG2SzRh?usp=drive_link)

*CODE BASE*

- [WhiteRabbit GitHub Repo](https://github.com/OHDSI/WhiteRabbit)

*OTHERS*

### General Description

**Mapping and Standardization** refers to making connections between source representations of medical events or concepts  (e.g. EPIC procedural code referring to an appendectomy) and standard representations of those elements (e.g. the [SNOMED procedural code for an appendectomy](https://athena.ohdsi.org/search-terms/terms/4198190)). Through the DelPhi process, we have generated a prioritized list of medical concepts that are relevant for the downstream analyses proposed in Bridge2AI. You can find this list, and multiple discussions of its interpretation and value, in the relevant resources section below.

### Expected Outputs

- Initial set of validated mappings that link source events to standard concepts in a tabular (ideally, `source_to_concept_map`) format that can be used in the transformation processes in C below. 

### Relevant Resources

*OFFICE HOURS*

- [Delphi MIMIC](https://drive.google.com/drive/folders/16ZtT6FATdpjhciXo8rkE_A6OaLLYs_K6?usp=drive_link)
- [Mapping Principles](https://drive.google.com/drive/folders/1Nz7WbZFJTZjBfrwJ_EAJtAMHFL-fDoyL?usp=drive_link)
- [Clinical Validation of Mappings Pt 1](https://drive.google.com/drive/folders/17yJKmPUCQOU7vy9Iy8F5CZun8t-ubpj7?usp=drive_link)
- [Clinical Validation of Mappings Pt 2](https://drive.google.com/drive/folders/1Dk-VYtTTVoPfWKKnsfDGStHJOTD4sbex?usp=drive_link)
- [Usagi and STCM Demo](https://drive.google.com/drive/folders/1OMB4Zhm1fw4p6wpkAR5CNQ0SEHnD9WMe?usp=drive_link)
- [Vocabulary Gaps and Mapping Discussion](https://drive.google.com/drive/folders/1Nz7WbZFJTZjBfrwJ_EAJtAMHFL-fDoyL?usp=drive_link)
- [Flowsheet Mappings Pt 1](https://drive.google.com/drive/folders/1fGClvXBWjb6pUJfcHCp_zY-g2Ii-JPUQ?usp=drive_link)
- [Flowsheet Mappings Pt 2](https://drive.google.com/drive/folders/1e1AGKmRPjQ4Wc_CFvFOqvkdYR6l8ofSy?usp=drive_link)
- [Flowsheet Mappings Pt 3](https://drive.google.com/drive/folders/1X8Hqz_dhBU4W_5S-fnQLfJWeKF05oQGM?usp=drive_link)
- [OMOP Vocabularies Pt 1](https://drive.google.com/drive/folders/1iSl3ZZmZYq5i8wLoJq5S3NIrdh9NXHkX?usp=drive_link)
- [OMOP Vocabularies Pt 2](https://drive.google.com/drive/folders/1CJT272NOPhI8WgQZcegkZyLZgiprS-0o?usp=drive_link)

*CODE BASE*

- [Usagi GitHub Repo](https://github.com/OHDSI/Usagi)

*OTHERS*

- [DELPHI Mapping Files](https://drive.google.com/drive/folders/1_yIBSu2ZIEQEejeYP7VJs0GJf12DNmG7?usp=drive_link)
- [Reducing Mapping Workload Discussion](https://github.com/chorus-ai/StandardsModule/discussions/80)
- [Sharing Mapping Experience Discussion](https://github.com/chorus-ai/StandardsModule/discussions/82)
- [NOTE Mapping Review Discussion](https://github.com/chorus-ai/StandardsModule/discussions/84)
- [DELPHI Mappings Discussion](https://github.com/chorus-ai/StandardsModule/discussions/85)
- [Athena OMOP Vocabularies](https://athena.ohdsi.org/search-terms/start)

### General Description

**Transforming OMOP Data** refers to rearranging source relational data into the OMOP CDM (v5.4) format. This process requires semantic mapping, described in section B above, as well as structural mapping, in which source fields (e.g. procedureinstance) are linked to target OMOP fields (e.g. procedure_occurrence_date).

### Expected Outputs

- Database instance containing relational EHR data in standard OMOP tables

### Relevant Resources

*OFFICE HOURS*

- [WhiteRabbit](https://drive.google.com/drive/folders/1QoQsawCY-AorcpOmz5bgzZACewG2SzRh?usp=drive_link)
- [External Vendor Considerations](https://drive.google.com/drive/folders/1UJ-r_jePkAjIfdLc7UGTEquGV7-yaR3k?usp=drive_link)
- [EHR Data Standardization](https://drive.google.com/drive/folders/1BtygVgWPkHfWLqPRyOpaiL_HHZfUl_mG?usp=drive_link)
- [Sample ETL Process](https://drive.google.com/drive/folders/1Tkovv9jsFhvyb0S5zEJvUaky2Nx1wKjh?usp=drive_link)
- [Setting up OMOP DDL](https://drive.google.com/drive/folders/14r5D8XgvYQ9-Vs4kTV2tF8NqjRFcJiH7?usp=drive_link)
- [Unit Harmonization](https://drive.google.com/drive/folders/1TAZddxdzSYt9vCowezlLtKT4CaXQUAr3?usp=drive_link)

*CODE BASE*

- [MIMIC ETL](https://github.com/chorus-ai/MIMIC)
- [OMOP CDM DDLs](https://github.com/OHDSI/CommonDataModel)
- [Rabbit in a Hat](https://github.com/OHDSI/WhiteRabbit)

*OTHERS*

- [Sample ETL Designs](https://drive.google.com/drive/folders/1A212ldq685Milyh07DjfxRd-Ayt0LZ3B?usp=drive_link)
- [CHoRUS ETL and Mapping Guidelines](https://drive.google.com/drive/folders/1hqdDQLrsSbTcgdAHnwbB32zrwHsHMr_e?usp=drive_link)
- [OMOP CDM Guidelines](https://ohdsi.github.io/CommonDataModel/cdm54.html)

### General Description

**Transforming Unstructured Note Data** refers to applying Natural Language Processing (NLP) algorithms to free-text notes (e.g. *patient displays signs of nausea*) in order to extract discrete medical concepts (e.g. [SNOMED condition *Nausea*](https://athena.ohdsi.org/search-terms/terms/31967) that can then be used in downstream prediction and estimation modeling. This process is tied to C1 in that the raw notes should be used to populate the [NOTE](https://ohdsi.github.io/CommonDataModel/cdm54.html#NOTE) table in OMOP CDM, and the associated extracted concepts are used to populate the [NOTE_NLP](https://ohdsi.github.io/CommonDataModel/cdm54.html#NOTE_NLP) table. Once in the NOTE_NLP table, these concepts can be then added to appropriate domain-specific tables (e.g. CONDITION_OCCURRENCE, OBSERVATION) using confidence thresholds and other established conventions.

### Expected Outputs

- NOTE table populated with sets of raw notes, associated with patients and encounters
- NOTE_NLP table populated with features extracted from those notes, first using a standard OHNLP protocol or a tailored CHoRUS concept set

### Relevant Resources

*OFFICE HOURS*

- [*To be discussed at a future session...*]

*CODE BASE*

- [OHNLP](https://github.com/OHNLP/Backbone)
- [MedTagger](https://github.com/OHNLP/MedTagger)
- [CHoRUS OHNLP](https://github.com/chorus-ai/OHNLP_tool)

*OTHERS*

- [NOTE Mapping Review Discussion](https://github.com/chorus-ai/StandardsModule/discussions/84)

### General Description

**Transforming Imaging and Waveform Data** refers to parsing and aggregating relevant files according to conventions established within CHoRUS and within the [OHDSI community](https://link.springer.com/article/10.1007/s10278-024-00982-6). The procedures and protocols for standardizing and transforming these data are still under active discussion within CHoRUS, but you can find a preliminary example in the [CHoRUS MIMIC repository](https://github.com/chorus-ai/MIMIC).

### Expected Outputs

- Appropriately named and organized image and waveform files
- OMOP events (initially in the PROCEDURE_OCCURRENCE table) that link each file with an OMOP person and an OMOP encounter

### Relevant Resources

*OFFICE HOURS*

- [Imaging Modalities and SubModalities](https://drive.google.com/drive/folders/1ScZ1UMLcBECrwfCqScAGzh4gUo5tHNDz?usp=drive_link)
- [Private Tags](https://drive.google.com/drive/folders/1vi4HsK34YS6f09PhiAd8z0D9nVtE0CMb?usp=drive_link)

*CODE BASE*

- [MIMIC Image Parsing](https://github.com/chorus-ai/MIMIC/blob/fcb9b22ef13424f04a65e612eaf2ffe1d72485ea/etl/src/etl/mimic.py#L233)
- [MIMIC Waveform Parsing](https://github.com/chorus-ai/MIMIC/blob/fcb9b22ef13424f04a65e612eaf2ffe1d72485ea/etl/src/etl/mimic.py#L285)

*OTHERS*

- [MIMIC ID Linking Description](https://github.com/chorus-ai/MIMIC/blob/main/README.md)
- [ID Assignment Standards Discussion](https://github.com/chorus-ai/data_acq_SOP/discussions/12)
- [Imaging Discussion](https://github.com/chorus-ai/data_acq_SOP/discussions/11)
- [Waveform Discussion](https://github.com/chorus-ai/data_acq_SOP/discussions/8)

### General Description

**Data Quality Checks, Iterative Updates, and QC Results Submission** refers to the process of verifying the plausibility, completeness, and conformance of the OMOP output. Here, we will use established open-source tools (e.g. Achilles, DQD, Ares) to execute a series of validated checks and then produce an extract (i.e. AresIndex) that can be visualized and compared with other OMOP instances with regard to its richness, quality, and diversity. It is this extract that data contributing sites will be required to submit to the central MGH cloud instance for evaluation and feedback. The quality checks and associated AresIndex extract form the basis of an iterative process of updating mappings and transformation logic in order to maintain as much granularity of the source data in OMOP format. 

### Expected Outputs

- AresIndex extract in a [standard format](https://github.com/chorus-ai/StandardsModule/discussions/86#discussioncomment-8411235)
- Successful upload of extract to central location
- Revisions to existing mappings or ETL logic

### Relevant Resources

*OFFICE HOURS*

- [Achilles Output](https://drive.google.com/drive/folders/1b2EhWZ08OvEjFj5A0uCAbR0SsbQan7e6?usp=drive_link)
- [DQD Output](https://drive.google.com/drive/folders/134ZQEQEOQCbjjFRHqHoxI-gDdsZJ9Utl?usp=drive_link)
- [DQD Overview](https://drive.google.com/drive/folders/16Z8ssUC0kolGJzjWqe2jd5ONFVef2neL?usp=drive_link)
- [Ares Overview](https://drive.google.com/drive/folders/19gv_n-WFqoFAM-1MHNpRa5heEUbrjoOg?usp=drive_link)
- [Contributing Ares Results Pt 1](https://drive.google.com/drive/folders/1LAmtp7kygnqr1vuGvKS3wmTvbn5gmNZ_?usp=drive_link)

*CODE BASE*

- [Achilles](https://github.com/OHDSI/Achilles)
- [DataQualityDashboard](https://github.com/OHDSI/DataQualityDashboard)
- [AresIndexer](https://github.com/OHDSI/AresIndexer)
- [Ares Web Application](https://github.com/OHDSI/Ares)

*OTHERS*

- [DQD Discussion](https://github.com/chorus-ai/StandardsModule/discussions/86)

### General Description

**Deidentification** refers to applying transformations to the data in OMOP CDM format that remove potentially identifying information. These processes can include elements like date shifting, obfuscation of source medical record numbers (MRNs), exclusion of highly sensitive health information (e.g. HIV-related events), and age-group consolidation. The exact specifications for how to apply these deidentification processes to source data will depend on (1) site-specific data sharing agreements, and (2) conventions established by CHoRUS.

### Expected Outputs

- An OMOP (+ multimodal) dataset that conforms to data sharing agreements and that is consistent in date shifting and patient obfuscation across data modalities

### Relevant Resources

*OFFICE HOURS*

- [*To be discussed at a future session...*]

*CODE BASE*

- [CHoRUS Imaging DeID](https://github.com/chorus-ai/chorus_imaging)
- [Privacy Scan Tool](https://github.com/chorus-ai/privacy_scan_tool)

*OTHERS*

### General Description

**Data Extraction** refers to placing data in the organizational structure defined by the CHoRUS DataAcquisition team. Thus far, the convention is to create per-person directories, each with three sub-directories (OMOP, Image, Waveform). This structure is subject to change depending on results of preliminary ingestion processes in the central cloud instance.

### Expected Outputs

- Directory structure with all data forms linked and organized per-person

### Relevant Resources

*OFFICE HOURS*

- [*To be discussed at a future session...*]

*CODE BASE*

- [MIMIC Image Parsing](https://github.com/chorus-ai/MIMIC/blob/fcb9b22ef13424f04a65e612eaf2ffe1d72485ea/etl/src/etl/mimic.py#L233)
- [MIMIC Waveform Parsing](https://github.com/chorus-ai/MIMIC/blob/fcb9b22ef13424f04a65e612eaf2ffe1d72485ea/etl/src/etl/mimic.py#L285)

*OTHERS*

- [Description of Data Extract Structure](https://chorus-ai.github.io/data_acq_SOP/docs/Data-Uploading/Data_uploading)
- [ID Assignment Standards Discussion](https://github.com/chorus-ai/data_acq_SOP/discussions/12)

### General Description

**Data Delivery** refers to the process of sharing the organized data extract with the central Azure instance hosted by MGH. This process is currently being done using [Azure Data Share](https://learn.microsoft.com/en-us/azure/data-share/overview) and the [Azure CLI](https://learn.microsoft.com/en-us/cli/azure/).

### Expected Outputs

- Data delivered

### Relevant Resources

*OFFICE HOURS*

- [*To be discussed at a future session...*]

*CODE BASE*

- [Central Data Ingestion](https://github.com/chorus-ai/chorus-container-apps/blob/main/etl/src/etl/ingest.py)
- [Central Data Merge](https://github.com/chorus-ai/chorus-container-apps/blob/main/etl/src/etl/merge.py)

*OTHERS*

- [Description of Data Extract Structure](https://chorus-ai.github.io/data_acq_SOP/docs/Data-Uploading/Data_uploading)
- [ID Assignment Standards Discussion](https://github.com/chorus-ai/data_acq_SOP/discussions/12)

**Comment by briangow** [comment]
@jshoughtaling , this is great!

**Comment by jshoughtaling** [comment]
This information has been superseded by the [CHoRUS SOP](https://chorus-ai.github.io/Chorus_SOP/) documentation site.
